text/0000-or-patterns.md

```diff
@@ -0,0 +1,666 @@
+       ),
+       (&Tilde, &(M(_) | MM(_, _))) => {
+           vr.constrain_lt(mmp.inclast());
+           vr.constrain_ge(mmp);
+       }
+       (&Compatible, &(MMP(0, minor, _) | MM(0, minor))) => {
+           vr.constrain_lt(MM(0, minor + 1));
+           vr.constrain_ge(mmp);
+       }
+       (&Compatible, &(MMP(major, _, _) | MM(major, _) | M(major))) => {
+           vr.constrain_lt(M(major + 1));
+           vr.constrain_ge(mmp);
+       }
+       ...,
+   }
+   ```
+
+8. From rustc, we have:
+
+   In `src/librustc_mir/interpret/eval_context.rs`:
+   ```rust
+   Some(Def::Static(..) | Def::Const(..) | Def::AssociatedConst(..)) => {},
+   ```
+
+   In `src/librustc_mir/util/borrowck_errors.rs`:
+   ```rust
+   (&ty::TyArray(_, _), Some(true) | None) => "array",
+   ```
+
+   In `src/librustc/middle/reachable.rs`:
+   ```rust
+   Some(Def::Local(node_id) | Def::Upvar(node_id, ..)) => { .. }
+   ```
+
+   In `src/librustc/infer/error_reporting/mod.rs`:
+   ```rust
+   Some(hir_map::NodeBlock(_) | hir_map::NodeExpr(_)) => "body",
+   ```
+
+   In `src/libfmt_macros/lib.rs`:
+   ```rust
+   Some((_, '>' | '<' | '^')) => { .. }
+   ```
+
+   In `src/librustc/traits/select.rs`:
+   ```rust
+   ty::TyInfer(ty::IntVar(_) | ty::FloatVar(_)) | .. => { .. }
+   ```
+
+   In `src/librustc_typeck/check/mod.rs`:
+   ```rust
+   ty::TyInt(ast::IntTy::I8 | ast::IntTy::I16) | ty::TyBool => { .. }
+
+   ...
+
+   ty::TyUint(ast::UintTy::U8 | ast::UintTy::U16) => { .. }
+   ```
+
+   In `src/tools/cargo/src/cargo/sources/path.rs`:
+   ```rust
+   Some("Cargo.lock" | "target") => continue,
+   ```
+
+   In `src/libsyntax_ext/format_foreign.rs`:
+   ```rust   
+   ('h' | 'l' | 'L' | 'z' | 'j' | 't' | 'q', _) => {
+       state = Type;
+       length = Some(at.slice_between(next).unwrap());
+       move_to!(next);
+   },
+
+   ...
+
+   let width = match self.width {
+       Some(Num::Next) => {
+           // NOTE: Rust doesn't support this.
+           return None;
+       }
+       w @ Some(Num::Arg(_) | Num::Num(_)) => w,
+       None => None,
+   };
+   ```
+
+   In `src/libsyntax/parse/token.rs`:
+
+   ```rust
+   BinOp(Minus | Star | Or | And) | OrOr => true,
+   ```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Simply put, `$p | $q` where `$p` and `$q` are some patterns is now itself
+a legal pattern.
+
+This means that you may for example write:
+
+```rust
+enum Foo<T> {
+    Bar,
+    Baz,
+    Quux(T),
+}
+
+fn main() {
+    match Some(Foo::Bar) {
+        Some(Foo::Bar | Foo::Baz) => { .. },
+        _ => { .. },
+    }
+}
+```
+
+Because `$p | $q` is itself a pattern, this means that you can nest arbitrarily:
+
+```rust
+fn main() {
+    match Some(Foo::Bar) {
+        Some(Foo::Bar | Foo::Quux(0 | 1 | 3)) => { .. },
+        _ => { .. }
+    }
+}
+```
+
+Note that the operator `|` has a low precedence. This means that you
+have to write `foo @ (1 | 2 | 3)` instead of writing `foo @ 1 | 2 | 3`.
```

Comment:
This phrasing seems confusing; "you have to write" suggests that this is what you'd want in the *common* case, in which case let's set the precedence to make that case easier to write.